src/Chapter2.hs

```diff
@@ -727,7 +770,12 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = go l []
+  where
+    go :: [Int] -> [Int] -> [Int]
+    go [] acc = acc
+    go (x:xs) acc = acc ++ replicate x x ++ go xs acc
```

Comment:
`acc` is always an empty list in this implementation. So you can safely remove this variable :slightly_smiling_face: When writing functions using the "Recursive go" pattern, you don't always need an accumulator. Sometimes you need one extra variables, sometimes two, sometimes none. But if it is not used, you are not forced to have it.

You are correct that after using `map` and `replicate` you have a list of lists. But you can use the `concat` function to flatten the list:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

The pattern of using `concat` and `map` is so common, so there's even a separate function `concatMap` which combines both (as the name suggests). After using `concatMap` and `replicate` you will get a pretty nice one-liner :slightly_smiling_face:

Reply:
Nice, that's super interesting! Thanks!